plaid/src/apis/npm/npm_web_client.rs

```diff
@@ -0,0 +1,833 @@
+            .await
+            .map_err(|_| ApiError::NpmError(NpmError::LoginFlowError))?;
+        let response = self
+            .client
+            .get(format!(
+                "{}/settings/{}/tokens",
+                NPMJS_COM_URL, self.config.username
+            ))
+            .header("X-Spiferack", "1") // to get JSON instead of HTML
+            .send()
+            .await
+            .map_err(|_| ApiError::NpmError(NpmError::FailedToListGranularTokens))?;
+
+        // Tokens are returned in a JSON structure under "list > objects". We first deserialize to a
+        // generic Value, then get "list > objects" and convert to a vector of NpmTokens. Finally, we
+        // keep only granular tokens and return the result.
+        let granular_tokens = serde_json::from_value::<Vec<NpmToken>>(
+            response
+                .json::<Value>()
+                .await
+                .map_err(|_| ApiError::NpmError(NpmError::FailedToListGranularTokens))?
+                .get("list")
+                .ok_or(ApiError::NpmError(NpmError::FailedToListGranularTokens))?
+                .get("objects")
+                .ok_or(ApiError::NpmError(NpmError::FailedToListGranularTokens))?
+                .clone(),
+        )
+        .map_err(|_| ApiError::NpmError(NpmError::FailedToListGranularTokens))?
+        .iter()
+        .filter(|v| v.token_type == Some("granular".to_string()))
+        .cloned()
+        .collect::<Vec<NpmToken>>();
+
+        serde_json::to_string(&granular_tokens)
+            .map_err(|_| ApiError::NpmError(NpmError::FailedToListGranularTokens))
+    }
+
+    /// Delete a package from the npm registry.
+    ///
+    /// Note: The package name should be unscoped. If you are trying to delete
+    /// @scope/package_name, then you should pass only "package_name". The scope is
+    /// preconfigured in the client and will be added automatically.
+    pub async fn delete_package(&self, package: &str, module: &str) -> Result<i32, ApiError> {
+        let package = self.validate_npm_package_name(package)?;
+        info!("Deleting npm package [{package}] on behalf of [{module}]");
+        self.login()
+            .await
+            .map_err(|_| ApiError::NpmError(NpmError::LoginFlowError))?;
+        // Step 1. Make a GET request to /delete in order to retrieve the dsrManifestHash.
+        // This will later be sent as form data when actually performing the deletion.
+        let response = self
+            .client
+            .get(format!(
+                "{}/package/%40{}%2F{}/delete",
+                NPMJS_COM_URL, self.config.npm_scope, package
+            ))
+            .send()
+            .await
+            .map_err(|_| ApiError::NpmError(NpmError::FailedToDeletePackage))?;
+        // Safe unwrap: hardcoded data.
+        // TODO best to precompile regex? This is likely an infrequent operation.
+        let dsr_manifest_hash = Regex::new(r#"dsrManifestHash\"\s+?value=\"([a-z0-9]{64})\""#)
```

Comment:
This is recompiling the regex every time and should be a validator

Reply:
Done: I made it a "special" validator because it's not sufficient to verify there is a match, but we also need to extract the token from the capturing group. Anyway, we are no longer recompiling the regex.